WordPress/gutenberg
# Update Navigation block to render hooked inner blocks

Fixes https://core.trac.wordpress.org/ticket/59743.

## What?


In this PR we are hooking into two filters. One which allows us to update the API response based on post type (`wp_navigation`) to insert hooked blocks before it gets returned to the editor. The second allows us to insert additional hooked blocks to the `core/navigation` via the `WP_Navigation_Block_Renderer` class before rendering happens on the frontend.

### Notes

This solution is very specific to the Navigation block and is not scalable when it comes to implementing it for other blocks. From my early understanding this is necessary since the Navigation block works slightly differently:

* The contents of the Navigation block is saved to the database as its own entity in `wp_posts` with a `post_type` of `wp_navigation`.
* The Navigation block hijacks the server-side rendering process in a very explicit way using `WP_Navigation_Block_Renderer` to render it on the server. I _believe_ this is due to a number of reasons but one of them being that it contains logic to wrap certain inner blocks in `<nav>` and `<li>` tags.
* This approach passes the current block as the `$context` value to the hook in addition to templates and patterns. You can differentiate this by checking that `$context['blockName']` is set to determine if it's a block.


### Todo

- [x] Insert hooked `first_child` and `last_child` blocks as inner blocks into the core Navigation block.
- [x] Investigate how this impacts fallback navigations / saved navigations.
- [x] Ensure that if a hooked block is removed from the core Navigation block that it remains removed.
- [x] Refactor code to reuse some utility functions used to insert hooked blocks into templates/parts/patterns.
- [x] Bypass unnecessary serialization/deserialization that is currently happening in the code.

Related: https://github.com/woocommerce/woocommerce/pull/43498

## Why?


## How?


## Testing Instructions
1. Add the below code to your themes `functions.php`
2. Load frontend and check the Login/Logout block is an inner block of the core Navigation block and hasn't been added twice.
3. Load the Header template part in the Site Editor and check that the Login/Logout block is present as an inner block of the core Navigation and hasn't been added twice.
4. Make customisations to move the block and check it persists between reloads.
5. Make customisations to remove the block completely and check it persists between reloads.
6. Remove the PHP code added in step 1, and now add the below JSON to the same blocks `block.json` file and retest starting from step 2.

```php
function register_logout_block_as_navigation_last_child( $hooked_blocks, $position, $anchor_block, $context ) {
	if ( $anchor_block === 'core/navigation' && $position === 'last_child' ) {
		$hooked_blocks[] = 'core/loginout';
	}

	return $hooked_blocks;
}

add_filter( 'hooked_block_types', 'register_logout_block_as_navigation_last_child', 10, 4 );
```

```json
"blockHooks": {
	"core/navigation": "lastChild"
}
```
